### DISCUSSION TITLE
API question plugin->host requests

### DISCUSSION THREAD

**Discussion by dbadb** [free-audio/clap#332] [Q&A] [answered]
Greetings clappers, 

I'm developing a CLAP host that hosts multiple plugins and am finding an awkward corner of Clap's plugin-to-host API. 

The problem: 

```c++
//host.h
request_callback(const clap_host *);
request_process(const clap_host *);
request_restart(const clap_host *);
// ext/gui.h (host)
request_resize(const clap_host_t *, uint32_t , uint32_t );
request_show(const clap_host_t *);
request_hide(const clap_host_t *);
```
These methods are invoked from a plugin to request a service from the host.  The `clap_host` struct includes a `host_data` field which I've been using (following the clap-host example) to stash a pointer to the C++ clap host instance.

The awkwardness is that these requests don't include any plugin-specific context.  Ie: the source of the request.   The only solution I see is to pass a plugin-instance-specific  clap_host with a different  host_data for each instance.   Obviously doable, but it would seem nicer if these three callbacks also included the `clap_plugin *` of the originator.

Any comments or thoughts are welcome!  Am I missing something?

Thanks!
Dana

**Comment by defiantnerd** [comment, marked as answer]
You will need to create an object instance for every hosted plugin. This will also contain a pointer back to the plugin instance.
In the clap-wrapper I've introduced an `IHost` interface and derived  the actual wrapping object from it. The pointer to this interface is being provided as `clap_host_t*`.

https://github.com/free-audio/clap-wrapper/blob/f14d5b24806f1da080d1534199d2668ebd2f5e43/src/wrapasvst3.h#L87

This way I can always find the way back to the plugin and I also know which instance is being addressed.

You will have no success with pointing `clap_host_t` to the same object for all plugin instances, it is individual for each plugin.

**Reply by dbadb** [reply]
Indeed - that is the flavor of the course I am taking.   That said, I'm curious if the simple API enhancement proposed above would also address your scenarios? 

Thanks for taking the time to comment.

**Reply by defiantnerd** [reply]
I don't think so - but it would introduce the possibility of using a non-matching `clap_host_t*`and `clap_plugin_t*`.
And since it actually just two pointers in a `struct` you have to maintain, this is something you can burden the host with. 

Additionally, you will have to maintain some object for each plugin instance anyway, so just use that - it will contain the `clap_plugin_t*`, I guess.
